### PR TITLE
resolved: be more careful with weird links with low MTUs

### PR DESCRIPTION
**Pull request by poettering**
Apparently CAN links will show up in rtnetlink with very low MTUs. We
shouldn't consider them relevant if no IP is spoken over them, since
these MTUs are irrelevant for us then.

Hence, let's check if there's an address assigned to the link before
considering its MTU.

As additional safety net filter out MTUs smaller than the minimum DNS
packet size, too.

Finally, in case we don't find any suitable interface MTU, let's default
to 1500 as the generic Ethernet MTU.

Fixes: #19396

### CODE REVIEW

File: src/resolve/resolved-manager.c [hunk 1]
```diff
@@ -1097,18 +1097,26 @@ uint32_t manager_find_mtu(Manager *m) {
...
+        /* If we don't know on which link a DNS packet would be delivered, let's find the largest MTU that
+         * works on all interfaces we know of that have an IP address asociated */
 
         HASHMAP_FOREACH(l, m->links) {
-                if (l->mtu <= 0)
+                /* Let's filter out links without IP addresses (e.g. AF_CAN links and suchlike) */
+                if (!l->addresses)
+                        continue;
+
+                /* Safety check: MTU shorter than what we eed for the absolutely shortest DNS request? Then
```
Comment (bluca) [inline src/resolve/resolved-manager.c]: /eed/need/

File: src/resolve/resolved-manager.c [hunk 2]
```diff
@@ -1097,18 +1097,26 @@ uint32_t manager_find_mtu(Manager *m) {
...
+                 * let's ignore this. */
+                if (l->mtu < DNS_PACKET_HEADER_SIZE + UDP4_PACKET_HEADER_SIZE)
```
Comment (bluca) [inline src/resolve/resolved-manager.c]: do we need to check if the address is ipv6, and then compare with UDP6_PACKET_HEADER_SIZE?

Comment (poettering) [reply src/resolve/resolved-manager.c]: not really, we only care for the smallest packet size, and ipv4 packets are smaller. To make this clearer I added a MIN() expression now that encodes this in code.

**Comment by poettering** [comment]
force pushed a new version, addressing the issues above. ptal!

**Review by bluca — APPROVED** [review]
Lgtm

**Comment by keszybz** [comment]
LGTM too.

**Pull request merged**